Source/JavaScriptCore/runtime/JSCell.cpp

```diff
@@ -338,4 +338,33 @@ NEVER_INLINE NO_RETURN_DUE_TO_CRASH NOT_TAIL_CALLED void reportZappedCellAndCras
 }
 #endif // CPU(X86_64)
 
+#if USE(JSVALUE32_64)
+bool isLiveConcurrently(const HeapCell* cell)
+{
+    if (!cell)
+        return false;
+
+    if (!isCompilationThread() || !Options::useConcurrentJIT())
+        return true;
+
+    // This also prevents pointer with invalid ranges, such as 0x02
+    MarkedBlock* candidate = MarkedBlock::blockFor(cell);
+    if (!candidate)
+        return false;
+
+    if (cell->isPreciseAllocation())
+        return cell->preciseAllocation().isLive();
```

Comment:
This is not correct. If random precise-allocation-like value comes, then this will crash.

Reply:
I didn't get it, could you elaborate? This is the same code as in `HeapCell::isLive()`.

Reply:
> I didn't get it, could you elaborate? This is the same code as in HeapCell::isLive().
No. HeapCell::isLive is OK because it is called after ensuring that this is a valid HeapCell address. This function is not guaranteeing that.

Reply:
Alright, does it mean that I need to do something like `isPointerGCObjectJSCell`?
```
        if (pointer->isPreciseAllocation()) {
            auto* set = heap.objectSpace().preciseAllocationSet();
            ASSERT(set);
            if (set->isEmpty())
                return false;
            if (!set->isValidValue(pointer))
                return false;
            return set->contains(pointer);
        }
```
I tried that but the `set` I get is always `0x0`.

Reply:
I added guards to try to retrieve the structure IDs, WDYT?

Reply:
Got that from `HeapVerifier::validateJSCell`

Reply:
I added  `ASSERT(!cell->isZapped());` and `ASSERT(cell->structureID());`, is it enough to make sure the pointer is a cell, right?

Reply:
You cannot just lift code from `HeapVerifier`.  `HeapVerifier` runs while the world is stopped, and it does not have to deal with concurrency issues.

To implement this feature, you need to understand the GC architecture very well at a fundamental level and think of the concurrency issues.